### PR TITLE
adding `WithCustomTag` to a few libs in `contrib/`: `gorm/gorm.v1`, `confluentinc/confluent-kafka-go/kafka`, `database/sql`, `google.golang.org/grpc`

### LINKED ISSUE #1308: Request: example for custom tags for `contrib/gorm.io`
hi DataDog team,

`dd-trace-go` supports [`gorm@v1`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1/contrib/jinzhu/gorm) and [`gorm@v2`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1@v1.38.1/contrib/gorm.io/gorm.v1) and I'm comparing the DataDog instrumentation implementation for both versions, trying to understand what would be the impact of upgrading my services from `gorm@v1` to `gorm@v2`.

one thing I noticed is: instrumentation lib for `gorm@v1` comes with an option to add custom tag: [`WithCustomTag`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1/contrib/jinzhu/gorm#WithCustomTag), and I don't see an equivalent option in `gorm@v2` ([reference](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1@v1.38.1/contrib/gorm.io/gorm.v1))

would it be possible for you to share an example of how to add custom tags using `dd-trace-go` for `gorm@v2`?  if it's not supported yet, would it be possible to add to `dd-trace-go/contrib/gorm.io`?

Thanks in advance!

### PR DESCRIPTION
**Pull request by duxing**
support custom tags for a few libs in `contrib/`: 
 - `gorm.io/gorm.v1`
 - `confluentinc/confluent-kafka-go/kafka`
 - `database/sql`
 - `google.golang.org/grpc`

Fixes #1308

### CODE REVIEW

File: contrib/gorm.io/gorm.v1/option.go [hunk 1]
```diff
@@ -71,3 +74,19 @@ func WithErrorCheck(fn func(err error) bool) Option {
...
+func WithCustomTag(tag string, tagFn func(db *gorm.DB) interface{}) Option {
+	return func(cfg *config) {
+		if cfg.tagFns == nil {
+			cfg.tagFns = make(map[string]func(scope *gorm.Scope) interface{})
+		}
+
+		if tagFn != nil {
+			cfg.tagFns[tag] = tagFn
+		} else {
+			delete(cfg.tagFns, tag)
```
Comment (ajgajg1134) [inline contrib/gorm.io/gorm.v1/option.go:89]: I'm curious why you went with a slightly different approach here compared to the other `WithCustomTag` function? Do you foresee needing to be able to remove/delete tagFns after they've already been added?

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]
Hi duxing! Thanks for this contribution, it looks like a good idea to bring both these versions to feature parity. Just one small comment I've attached, and also can you also add a unit test for this change? You may be able to borrow nearly all the test from the previous version which has a test `TestCustomTags` that covers this.

Thanks again!

File: contrib/gorm.io/gorm.v1/option.go (hunk 1 above)
Comment (duxing) [reply contrib/gorm.io/gorm.v1/option.go:89]: tbh I don't think it's practically useful. added this implementation to behave in a way that callers would assume. Not a must-have and I can remove it if you prefer.

**Comment by duxing** [comment]
hi @ajgajg1134  sorry about the delay. I was able to get to this last night and added test coverage for everything I introduced. can you take another look when you have a chance?

I noticed the CI fails complaining about `DD_API_KEY` missing in the env (and another `bitnami/kafka` related issue). I'm not sure how to address these ones, it'll be helpful if you can share some insights with me

**Comment by ajgajg1134** [comment]
> hi @ajgajg1134 can you help me understand why my PR fails due to:
> 
> > Neither DATADOG_API_KEY nor DD_API_KEY is in your environment.
> > Internal Error: API key is missing
> > when other PRs (eg. [this](https://github.com/DataDog/dd-trace-go/pull/1348) and [this](https://github.com/DataDog/dd-trace-go/pull/1267) ) don't have the problem? it's probably something trivial and I'd like to figure out what I missed.
> 
> thanks in advance!

Hello @duxing! Sorry, the system-test failures you're seeing are due to this PR running a branch from an external fork of the repository. Unfortunately it means these tests can't be run unless the branch is coming from within the repository, although we can run the other tests still and should be able to merge after review is complete. (This is a limitation of github actions that hopefully is resolved at some point)

**Comment by duxing** [comment]
@ajgajg1134 thx for explaining, i figured it has something to do with the github action settings. I've disabled my own fork since it's complaining about the same thing.

is it possible to cherry-pick my change to a branch for this repo and proceed?

**Comment by ajgajg1134** [comment]
Closing this PR since the work was merged with #1359

**Pull request closed without merging**